File: src/codes/Users_Manual/advanced/custom_gain_0.rs
use std::collections::HashMap;
use autd3::prelude::*;
use autd3::derive::*;

#[derive(Gain)]
pub struct FocalPoint {
    position: Vector3,
}

impl FocalPoint {
    pub fn new(position: Vector3) -> Self {
        Self { position }
    }
}

impl Gain for FocalPoint {
    fn calc(&self, _geometry: &Geometry) -> GainCalcResult {
        let position = self.position;
        Ok(Self::transform(move |dev| {
            let wavenumber = dev.wavenumber();
            move |tr| {
                Drive::new(
                    Phase::from((position - tr.position()).norm() * wavenumber * rad),
                    EmitIntensity::MAX,
                )
            }
        }))
    }
}
# fn main() { 
# }

File: src/codes/Users_Manual/advanced/custom_gain_0.cpp
//~#include "autd3.hpp"
//~#include <unordered_map>
//~#include <limits>
class FocalPoint final : public autd3::derive::Gain<FocalPoint> {
 public:
  explicit FocalPoint(autd3::Vector3 point) : _point(std::move(point)) {}

  autd3::gain::GainCalcResult calc(
      const autd3::Geometry& geometry) const override {
    return transform([&](const auto& dev) {
      return [&](const auto& tr) {
        const auto phase = (tr.position() - _point).norm() * dev.wavelength();
        return autd3::Drive{autd3::Phase(phase * autd3::rad),
                            std::numeric_limits<autd3::EmitIntensity>::max()};
      };
    });
  }

 private:
  autd3::Vector3 _point;
};
//~int main() { return 0; }